Readme (#371)
feldera/dbsp: CONTRIBUTING.md

### Merging a pull request

> Since we run benchmarks as part of the CI, it'sa  good practice to preserve the commit IDs of the feature branch

Reviewer: ```suggestion
Since we run benchmarks as part of the CI, it's a good practice to preserve the commit IDs of the feature branch
```

git push upstream main
```

### Code Style

> Execute the following command to make `git push` check the code for formatting issues. It is not yet applied to the sql compiler.

Reviewer: It just runs `cargo fmt --check`. The SQL compiler is in Java. Do you have an equivalent command for Java?

> Execute the following command to make `git push` check the code for formatting issues. It is not yet applied to the sql compiler.

Reviewer 2: I think he means the runtime Rust library in lib

> Execute the following command to make `git push` check the code for formatting issues. It is not yet applied to the sql compiler.

Author: Yeah I mean the rust libraries in the sql folder.

> Execute the following command to make `git commit` check the code for formatting issues before commit. It is not yet applied to the sql compiler.

Reviewer: I think this checks in `git push` instead of `git commit`.

## Reporting Bugs and Creating Issues

When opening a new issue, try to roughly follow the commit message format conventions above.

# For developers

## Building DBSP from sources

DBSP is implemented in Rust and uses Rust's `cargo` build system. The SQL
to DBSP compiler is implemented in Java and uses `maven` as its build system.

You can build the rust sources by runnning the following at the top level of this tree.

```
cargo build
```

To build the SQL to DBSP compiler, run the following from `sql-to-dbsp-compiler/SQL-compiler`:

```
mvn package
```

If you want to develop DBSP without installing the required toolchains
locally, you can use Github Codespaces; from
https://github.com/feldera/dbsp, click on the green `<> Code` button,
then select Codespaces and click on "Create codespace on main".

## Learning the DBSP rust code

To learn how the DBSP core works, we recommend starting with the tutorial.

From the project root:

```
cargo doc --open
```

> Then search for `dbsp::tutorial`.

Reviewer 2: We should really publish the docs and link to them.
Maybe not now.
Can't you put here the exact location?

Author: I couldn't find a way to pass an exact argument to cargo doc :(

Each benchmark has its own options and behavior, as outlined below.

### Nexmark Benchmark

You can run the complete set of Nexmark queries, with the default settings, with:

```shell
cargo bench --bench nexmark --features with-nexmark
```

By default this will run each query with a total of 100 million events emitted at 10M per second (by two event generator threads), using 2 CPU cores for processing the data.

To run just the one query, q3, with only 10 million events, but using 8 CPU cores to process the data and 6 event generator threads, you can run:

```shell
cargo bench --bench nexmark --features with-nexmark -- --query q3 --max-events 10000000 --cpu-cores 8 --num-event-generators 6
```

For further options that you can use with the Nexmark benchmark,

```shell
cargo bench --bench nexmark --features with-nexmark -- --help
```

An extensive blog post about the implementation of Nexmark in DBSP:
<https://liveandletlearn.net/post/vmware-take-3-experience-with-rust-and-dbsp/>

## Updating the pipeline manager database schema

> Here are some guidelines when contributing code that affects the Pipeline Manager's DB schema.

Reviewer 2: This is rather sudden. A couple of phrases explaining what the pipeline manager is?

> * We use SQL migrations to apply the schema to a live database to faciliate upgrades. We use [refinery](https://github.com/rust-db/refinery) to manage migrations.

Reviewer: ```suggestion
* We use SQL migrations to apply the schema to a live database to facilitate upgrades. We use [refinery](https://github.com/rust-db/refinery) to manage migrations.
```

* The migration files can be found in `crates/pipeline_manager/migrations`
* Do not modify an existing migration file. If you want to evolve the schema, add a new SQL or rust file to the migrations folder following [refinery's versioning and naming scheme](https://docs.rs/refinery/latest/refinery/#usage). The migration script should update an existing schema as opposed to assuming a clean slate. For example, use `ALTER TABLE` to add a new column to an existing table and fill that column for existing rows with the appropriate defaults.

> * If you add a new migration script `V{i}`, add tests for migrations from `V{i-1} to V{i}`. For example, add tests that invoke the pipeline manager APIs before and after the migration.

Reviewer: ```suggestion
* If you add a new migration script `V{i}`, add tests for migrations from `V{i-1}` to `V{i}`. For example, add tests that invoke the pipeline manager APIs before and after the migration.
```
